### ISSUE TITLE
DateTimePicker React error maximum depth when setting the value state

### ISSUE THREAD

**Issue by Martijn-Schoenmaker-Webbio** [strapi/design-system#867]
### Required System information

- Node.js version: 16.19.0
- NPM version: 8.19.3
- Strapi version: 4.5.6
- Database: Mysql
- Operating system: Windows 11

### Describe the bug

The DateTimePicker creates a React error `Warning: Maximum update depth exceeded` when trying to set the value from a state.

### Steps to reproduce the behavior

Use the `DateTimePicker` from `'@strapi/design-system'` and set the value via another component

### Expected behavior

I would expect to be able to set the value of the component.

### Code snippets

```ts
import React, { useState } from "react";

import { DateTimePicker, Button,  Box } from "@strapi/design-system";

const Component = () => {
  const [date, setDate] = useState<Date | null>(null);

  return (
    <Box>
      <Button type="button" aria-label="Reset" onClick={() => setDate(new Date())}>
        Click me
      </Button>

      <DateTimePicker
        ariaLabel="Label"
        label="Label"
        name="date"
        onChange={(e: Date) => setDate(e)}
        size="S"
        value={date}
      />
    </Box>
  );
};
```

**Comment by joshuaellis** [comment]
Would you be able to share a codesandbox of the issue? :)

**Comment by Martijn-Schoenmaker-Webbio** [comment]
Not sure where to find the template sandbox, but found this one: https://codesandbox.io/s/objective-shockley-p1wi80

**Comment by Martijn-Schoenmaker-Webbio** [comment]
For what it's worth I created a date-time-picker based on Strapi's, that doesn't have the bug.

``` ts
import React, { useState, useEffect } from 'react';

import { DatePicker, TimePicker, Stack, Flex, FieldLabel, FieldHint, FieldError } from '@strapi/design-system';

import S from './styles';

export interface IDateTimePickerProps {
	ariaLabel: string;
	clearLabel?: string;
	disabled?: boolean;
	error?: string;
	hint?: string;
	label?: string;
	labelAction?: JSX.Element;
	onChange?: (date: Date | null) => void;
	onClear?: () => void;
	name?: string;
	required?: boolean;
	size?: string;
	step?: number;
	value?: Date | null;
}

const DateTimePicker = ({
	ariaLabel,
	clearLabel,
	disabled,
	error,
	hint,
	label,
	labelAction,
	onChange,
	onClear,
	name,
	required,
	size,
	step,
	value,
	...props
}: IDateTimePickerProps) => {
	const [dateValue, setDateValue] = useState<Date | null>(getDateFromValue(value));
	const [timeValue, setTimeValue] = useState<string | null>(getTimeValueFromDate(dateValue));

	useEffect(() => {
		setDateValue(getDateFromValue(value));
		setTimeValue(getTimeValueFromDate(value));
	}, [value]);

	const handleDateChange = (e: Date) => {
		const newDate = setDateWithTime(e, timeValue);

		setDateValue(newDate);
		onChange?.(newDate);
	};

	const handleTimeChange = (e: string) => {
		const newDate = setDateWithTime(dateValue, e);

		setDateValue(newDate);
		onChange?.(newDate);
	};

	const handleTimeClear = () => {
		handleTimeChange('00:00');
	};

	const handleDateClear = () => {
		setDateValue(null);
		setTimeValue(null);
		onClear?.();
	};

	return (
		<S.CustomField
			name={name}
			role="group"
			aria-labelledby="datetime-label"
			hint={hint}
			error={error}
			required={required}
		>
			<Stack spacing={1}>
				{label && (
					<Flex>
						<FieldLabel id="datetime-label">{label}</FieldLabel>
						{labelAction && <S.LabelAction paddingLeft={1}>{labelAction}</S.LabelAction>}
					</Flex>
				)}
				<Stack horizontal spacing={2}>
					<DatePicker
						name={name}
						ariaLabel={label || ariaLabel}
						error={error}
						selectedDate={dateValue}
						selectedDateLabel={(formattedDate) => `Date picker, current is ${formattedDate}`}
						onChange={handleDateChange}
						size={size}
						onClear={onClear && handleDateClear}
						clearLabel={clearLabel}
						disabled={disabled}
						{...props}
					/>
					<TimePicker
						size={size}
						aria-label={label || ariaLabel}
						error={error}
						value={timeValue}
						onChange={handleTimeChange}
						onClear={onClear && handleTimeClear}
						clearLabel={clearLabel}
						disabled={disabled}
						step={step}
					/>
				</Stack>
				<FieldHint />
				<FieldError id="datetimepicker" />
			</Stack>
		</S.CustomField>
	);
};

export default DateTimePicker;

DateTimePicker.defaultProps = {
	ariaLabel: undefined,
	clearLabel: 'clear',
	disabled: false,
	error: undefined,
	hint: undefined,
	label: undefined,
	labelAction: undefined,
	onClear: undefined,
	required: false,
	size: 'M',
	step: 1,
	value: undefined
};

const setDateWithTime = (date?: Date | null, time?: string | null): Date | null => {
	if (!date && !time) {
		return null;
	}

	const newDate = date ? new Date(date) : new Date();
	const times = getNumbersFromString(time);

	if (times) {
		newDate.setHours(times[0]);
		newDate.setMinutes(times[1]);
	}

	return newDate;
};

const getNumbersFromString = (timeString?: string | null): undefined | [number, number] => {
	if (!timeString) {
		return undefined;
	}

	const hours = Number(timeString.split(':')[0]);
	const minutes = Number(timeString.split(':')[1]);

	const arr = [hours, minutes].filter((v) => v || v === 0);

	return arr.length === 2 ? (arr as [number, number]) : undefined;
};

const getDateFromValue = (date?: Date | null): Date | null => {
	if (!date) {
		return null;
	}

	return date;
};

const getTimeValueFromDate = (date?: Date | null): string | null => {
	if (!date) {
		return null;
	}

	return [date.getHours(), date.getMinutes(), date.getSeconds()].join(':');
};
```

**Comment by joshuaellis** [comment]
> For what it's worth I created a date-time-picker based on Strapi's, that doesn't have the bug.
> 
> ```tsx
> // date-time-picker.tsx
> import React, { useState, useEffect } from 'react';
> 
> import { DatePicker, TimePicker, Stack, Flex, FieldLabel, FieldHint, FieldError } from '@strapi/design-system';
> 
> import S from './styles';
> 
> export interface IDateTimePickerProps {
> 	ariaLabel: string;
> 	clearLabel?: string;
> 	disabled?: boolean;
> 	error?: string;
> 	hint?: string;
> 	label?: string;
> 	labelAction?: JSX.Element;
> 	onChange?: (date: Date | null) => void;
> 	onClear?: () => void;
> 	name?: string;
> 	required?: boolean;
> 	size?: string;
> 	step?: number;
> 	value?: Date | null;
> }
> 
> const DateTimePicker = ({
> 	ariaLabel,
> 	clearLabel,
> 	disabled,
> 	error,
> 	hint,
> 	label,
> 	labelAction,
> 	onChange,
> 	onClear,
> 	name,
> 	required,
> 	size,
> 	step,
> 	value,
> 	...props
> }: IDateTimePickerProps) => {
> 	const [dateValue, setDateValue] = useState<Date | null>(getDateFromValue(value));
> 	const [timeValue, setTimeValue] = useState<string | null>(getTimeValueFromDate(dateValue));
> 
> 	useEffect(() => {
> 		setDateValue(getDateFromValue(value));
> 		setTimeValue(getTimeValueFromDate(value));
> 	}, [value]);
> 
> 	const handleDateChange = (e: Date) => {
> 		const newDate = setDateWithTime(e, timeValue);
> 
> 		setDateValue(newDate);
> 		onChange?.(newDate);
> 	};
> 
> 	const handleTimeChange = (e: string) => {
> 		const newDate = setDateWithTime(dateValue, e);
> 
> 		setDateValue(newDate);
> 		onChange?.(newDate);
> 	};
> 
> 	const handleTimeClear = () => {
> 		handleTimeChange('00:00');
> 	};
> 
> 	const handleDateClear = () => {
> 		setDateValue(null);
> 		setTimeValue(null);
> 		onClear?.();
> 	};
> 
> 	return (
> 		<S.CustomField
> 			name={name}
> 			role="group"
> 			aria-labelledby="datetime-label"
> 			hint={hint}
> 			error={error}
> 			required={required}
> 		>
> 			<Stack spacing={1}>
> 				{label && (
> 					<Flex>
> 						<FieldLabel id="datetime-label">{label}</FieldLabel>
> 						{labelAction && <S.LabelAction paddingLeft={1}>{labelAction}</S.LabelAction>}
> 					</Flex>
> 				)}
> 				<Stack horizontal spacing={2}>
> 					<DatePicker
> 						name={name}
> 						ariaLabel={label || ariaLabel}
> 						error={error}
> 						selectedDate={dateValue}
> 						selectedDateLabel={(formattedDate) => `Date picker, current is ${formattedDate}`}
> 						onChange={handleDateChange}
> 						size={size}
> 						onClear={onClear && handleDateClear}
> 						clearLabel={clearLabel}
> 						disabled={disabled}
> 						{...props}
> 					/>
> 					<TimePicker
> 						size={size}
> 						aria-label={label || ariaLabel}
> 						error={error}
> 						value={timeValue}
> 						onChange={handleTimeChange}
> 						onClear={onClear && handleTimeClear}
> 						clearLabel={clearLabel}
> 						disabled={disabled}
> 						step={step}
> 					/>
> 				</Stack>
> 				<FieldHint />
> 				<FieldError id="datetimepicker" />
> 			</Stack>
> 		</S.CustomField>
> 	);
> };
> 
> export default DateTimePicker;
> 
> DateTimePicker.defaultProps = {
> 	ariaLabel: undefined,
> 	clearLabel: 'clear',
> 	disabled: false,
> 	error: undefined,
> 	hint: undefined,
> 	label: undefined,
> 	labelAction: undefined,
> 	onClear: undefined,
> 	required: false,
> 	size: 'M',
> 	step: 1,
> 	value: undefined
> };
> 
> const setDateWithTime = (date?: Date | null, time?: string | null): Date | null => {
> 	if (!date && !time) {
> 		return null;
> 	}
> 
> 	const newDate = date ? new Date(date) : new Date();
> 	const times = getNumbersFromString(time);
> 
> 	if (times) {
> 		newDate.setHours(times[0]);
> 		newDate.setMinutes(times[1]);
> 	}
> 
> 	return newDate;
> };
> 
> const getNumbersFromString = (timeString?: string | null): undefined | [number, number] => {
> 	if (!timeString) {
> 		return undefined;
> 	}
> 
> 	const hours = Number(timeString.split(':')[0]);
> 	const minutes = Number(timeString.split(':')[1]);
> 
> 	const arr = [hours, minutes].filter((v) => v || v === 0);
> 
> 	return arr.length === 2 ? (arr as [number, number]) : undefined;
> };
> 
> const getDateFromValue = (date?: Date | null): Date | null => {
> 	if (!date) {
> 		return null;
> 	}
> 
> 	return date;
> };
> 
> const getTimeValueFromDate = (date?: Date | null): string | null => {
> 	if (!date) {
> 		return null;
> 	}
> 
> 	return [date.getHours(), date.getMinutes(), date.getSeconds()].join(':');
> };
> ```
> 
> ```ts
> // styles.ts
> 
> import styled from 'styled-components';
> 
> import { Box, Field } from '@strapi/design-system';
> 
> const CustomField = styled(Field)`
> 	> div {
> 		> div {
> 			p[data-strapi-field-error='true'] {
> 				display: none;
> 			}
> 		}
> 	}
> `;
> 
> const LabelAction = styled(Box)`
> 	svg path {
> 		fill: ${({ theme }) => theme.colors.neutral500};
> 	}
> `;
> 
> const DateTimePickerStyles = {
> 	CustomField,
> 	LabelAction
> };
> 
> export default DateTimePickerStyles;
> ```

Feel free to submit a PR for this, be great to see it in detail :)

**Comment by eucciferri** [comment]
Hi @Martijn-Schoenmaker-Webbio! I was wondering to know if you were able to submit a PR like @joshuaellis said in the comment above. Thanks in advance!!

**Comment by Martijn-Schoenmaker-Webbio** [comment]
I would create a PR, but my files are TS files. I don't know if you will convert the codebase from JS to TS?

**Comment by joshuaellis** [comment]
> I would create a PR, but my files are TS files. I don't know if you will convert the codebase from JS to TS?

If you want to convert the file to TS, I will be v happy <3

**Comment by Martijn-Schoenmaker-Webbio** [comment]
If i try to run the repo locally I get a ton of warnings. Can you explain how to test this repo locally so I can create a pr?

**Comment by joshuaellis** [comment]
Warnings are a bit ambiguous, does it install if you run yarn setup and then yarn develop?

**Comment by Martijn-Schoenmaker-Webbio** [comment]
Setup crashes on `Entry module cannot be external (src/index.tsx) --> @strapi/icons:build`

**Comment by ntarandek** [comment]
This small change resolved the issue. Removed the value !== dateValue check as well as dateValue from dependency array. I managed to replicate the issue in story book by changing the story - so this really works.

Can someone check this to confirm the fix and prepare a PR?
Seems that some people are waiting for this fix - and we are stuck on Strapi 4.5.4 because of this.

```
useEffect(() => {
    if (value) {
      const parsedData = parseDate(value);
      setDateValue(parsedData);
      setTimeValue(
        parsedData ? `${parsedData.getHours()}:${parsedData.getMinutes()}:${parsedData.getSeconds()}` : null,
      );
    } else if (!value) {
      setDateValue(undefined);
      setTimeValue(undefined);
    }
  }, [value]);
```

Change in story to check the error - just added a button to set the value programmatically, which starts the issue.

```
<Canvas>
  <Story name="base">
    {() => {
      const [value, setValue] = useState(undefined);
      return (
        <>
        <Button variant="danger" onClick={() => setValue(new Date())}>
          Replicate error
        </Button>
        <br/>
        <DateTimePicker
          label={'Date time picker'}
          disabled={false}
          name="datetimepicker"
          onChange={(e) => setValue(e)}
          clearLabel="Clear"
          onClear={() => setValue(undefined)}
          value={value}
          hint="This is a super description"
          selectedDateLabel={() => `Date picker, current is undefined`}
        />
        </>
      );
    }}
  </Story>
</Canvas>
```

**Comment by eucciferri** [comment]
Hello guys!
Does anyone know if this issue was solved?
Thanks!!

**Comment by ntarandek** [comment]
@eucciferri It seems that we simply have to wait since it obviously takes a long while to get some replies here. Maybe we are the only ones affected (which I doubt), so there is not a big incentive to resolve this quickly.  But even if that is the case, my 2 weeks old solution seems quite easy to implement, but I do not want to presume (for now) what the delay is all about.

**Comment by joshuaellis** [comment]
Hi everyone, thank you for your patience I appreciate this is frustrating.

I've begun working on this today. It'll either be released in `1.6.5` or `1.6.6`. From there I can give you a better idea of what version of the strapi admin panel it'll be fixed in.

**Comment by claudiograssia** [comment]
Hi guys,
This bug was resolved?

Thanks

**Comment by joshuaellis** [comment]
> Hi guys, This bug was resolved?
> 
> Thanks

Nope, there's a PR we're working on that solves the issue however :)

**Comment by joshuaellis** [comment]
@eucciferri, you have a few choices:

1. use patch package (or maybe yarn resolutions) to change all your `@strapi/design-system` dependencies to `1.6.6`
2. keep an eye on this PR https://github.com/strapi/strapi/pull/16113 & there should be an experimental release the next day after it's merged (it's on a CRON job)
3. It'll be released in `4.8.3` which _should_ be released on the 29th March afaik.
